src/ci_ex_states/excited_state_solver.cc

```diff
@@ -0,0 +1,1131 @@
+    if (sparse_solver_->sigma_method_ == "HZ") {
+        op.clear_op_s_lists();
+        op.clear_tp_s_lists();
+        op.build_strings(space);
+        op.op_lists(space);
+        op.tp_lists(space);
+    }
+
+    if (diag_method_ == Dynamic) {
+        for (size_t n = 0; n < nroot_; ++n) {
+            double S2 = op.s2_direct(space, evecs, n);
+            double S = std::fabs(0.5 * (std::sqrt(1.0 + 4.0 * S2) - 1.0));
+            spin_vec[n] = std::make_pair(S, S2);
+        }
+    } else {
+        for (size_t n = 0; n < nroot_; ++n) {
+            double S2 = op.s2(space, evecs, n);
+            double S = std::fabs(0.5 * (std::sqrt(1.0 + 4.0 * S2) - 1.0));
+            spin_vec[n] = std::make_pair(S, S2);
+        }
+    }
+    return spin_vec;
+}
+
+double ExcitedStateSolver::compute_spin_contamination(DeterminantHashVec& space, WFNOperator& op,
+                                                      psi::SharedMatrix evecs, int nroot) {
+    auto spins = compute_spin(space, op, evecs, nroot);
+    double spin_contam = 0.0;
+    for (int n = 0; n < nroot; ++n) {
+        spin_contam += spins[n].second;
+    }
+    spin_contam /= static_cast<double>(nroot);
+    spin_contam -= (0.25 * (state_.multiplicity() * state_.multiplicity() - 1.0));
+
+    return spin_contam;
+}
+
+std::vector<RDMs> ExcitedStateSolver::rdms(const std::vector<std::pair<size_t, size_t>>& root_list,
+                                           int max_rdm_level) {
+
+    std::vector<RDMs> refs;
+
+    for (const auto& root_pair : root_list) {
+
+        refs.push_back(compute_rdms(as_ints_, final_wfn_, op_, evecs_, root_pair.first,
+                                    root_pair.second, max_rdm_level));
+    }
+    return refs;
+}
+
+std::vector<RDMs>
+ExcitedStateSolver::transition_rdms(const std::vector<std::pair<size_t, size_t>>& root_list,
+                                    std::shared_ptr<ActiveSpaceMethod> method2, int max_rdm_level) {
+    std::vector<RDMs> refs;
+    throw std::runtime_error("ExcitedStateSolver::transition_rdms is not implemented!");
+    return refs;
+}
+
+RDMs ExcitedStateSolver::compute_rdms(std::shared_ptr<ActiveSpaceIntegrals> fci_ints,
+                                      DeterminantHashVec& dets, WFNOperator& op,
+                                      psi::SharedMatrix& PQ_evecs, int root1, int root2,
+                                      int max_rdm_level) {
+
+    if (!direct_rdms_) {
+        op.clear_op_s_lists();
+        op.clear_tp_s_lists();
+        if (diag_method_ == Dynamic) {
+            op.build_strings(dets);
+        }
+        op.op_s_lists(dets);
+        op.tp_s_lists(dets);
+
+        if (max_rdm_level >= 3) {
+            psi::outfile->Printf("\n  Computing 3-list...    ");
+            local_timer l3;
+            op_.three_s_lists(final_wfn_);
+            psi::outfile->Printf(" done (%1.5f s)", l3.get());
+        }
+    }
+
+    CI_RDMS ci_rdms_(dets, fci_ints, PQ_evecs, root1, root2);
```

Comment:
This one is my fault from a long time ago, this variable name should be `ci_rdms` without the final underscore